cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -0,0 +1,134 @@
+import io.cdap.cdap.sourcecontrol.UnexpectedRepositoryChangesException;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+
+/**
+ * In-Memory implementation for {@link SourceControlOperationRunner}.
+ * Runs all git operation inside calling service.
+ */
+public class InMemorySourceControlOperationRunner implements SourceControlOperationRunner {
+  private final RepositoryManagerFactory repoManagerFactory;
+  private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+  private static final Logger LOG = LoggerFactory.getLogger(InMemorySourceControlOperationRunner.class);
+
+  @Inject
+  public InMemorySourceControlOperationRunner(RepositoryManagerFactory repoManagerFactory) {
+    this.repoManagerFactory = repoManagerFactory;
+  }
+
+  @Override
+  public PushAppResponse push(NamespaceId namespace, RepositoryConfig repoConfig,
+                              ApplicationDetail appToPush, CommitMeta commitDetails)
+    throws Exception {
+    try (RepositoryManager repositoryManager = repoManagerFactory.create(namespace, repoConfig)) {
+      try {
+        repositoryManager.cloneRemote();
+      } catch (Exception e) {
+        throw new PushFailureException("Failed to clone remote repository", e);
+      }
+      LOG.info("Pushing application configs for : {}", appToPush);
+
+      //TODO CDAP-20371: Add retry logic here in case the head at remote moved while we are doing push
+      String pushedCommit = writeAppDetailAndPush(repositoryManager, appToPush, commitDetails);
+
+      try {
+        Path relativeFilePath = getAndValidateAppConfigRelativePath(repositoryManager, appToPush.getName());
+        String gitFileHash = repositoryManager.getFileHash(relativeFilePath, pushedCommit);
+        return new PushAppResponse(appToPush.getName(), appToPush.getAppVersion(), gitFileHash);
+      } catch (Exception e) {
+        throw new PushFailureException("Failed to fetch push details", e);
+      }
+    }
+  }
+
+  private String writeAppDetailAndPush(RepositoryManager repositoryManager,
+                                       ApplicationDetail appToPush,
+                                       CommitMeta commitDetails)
+    throws PushFailureException, NoChangesToPushException {
+    Path appWrittenPath;
+    try {
+      String appName = appToPush.getName();
+      // Symlinks will be followed but if the path points to an outside directory it will fail
+      appWrittenPath = getAndValidateAppConfigRelativePath(repositoryManager, appName);
+      // Opens the file for writing, creating the file if it doesn't exist,
+      // or truncating an existing regular-file to a size of 0
+      Path filePathToWrite = repositoryManager.getRepositoryRoot().resolve(appWrittenPath);
```

Comment:
we're getting the full path then relativizing it then resolving it again here. 

Instead, generate the relative path once before validating. Change the method above to validate and return the absolute path, then use that here.

Reply:
So we're validating & generating the relative path because the `getFileHash` method is taking a relative Path. It was sugguested by @tivv . The context is: https://github.com/cdapio/cdap/pull/14917#discussion_r1106272625

Reply:
that's fine, that's why I said to get the relative path first, pass it to validate(), then use it down below. In other words, take out some of the logic in getAndValidateAppConfigRelativePath() so we're not converting back and forth for no reason.

Reply:
got it. Now we get the relative path and validate it, return the absolute path for writing the configs